lib/history/sqlite/event.go

```diff
@@ -106,240 +77,184 @@ type sqlExecer struct {
-	return clusterRecovered{
-		sqlEvent: sqlEvent{
-			Timestamp: timestamp,
-			EventType: clusterRecoveredType,
-		},
-	}
-}
-
-// Insert inserts event into storage using the provided execer.
-func (e clusterRecovered) Insert(ctx context.Context, execer history.Execer) error {
+func (r *clusterRecovered) Insert(ctx context.Context, execer history.Execer) error {
 	const insertStmt = "INSERT INTO events (timestamp, type) VALUES (?,?)"
-	args := []interface{}{e.Timestamp, e.EventType}
+	args := []interface{}{r.GetTimestamp().ToTime(), history.ClusterRecovered}
 	return trace.Wrap(execer.Exec(ctx, insertStmt, args...))
 }
 
-// ProtoBuf returns event as a protobuf message.
-func (e clusterRecovered) ProtoBuf() (*pb.TimelineEvent, error) {
-	return history.NewClusterRecovered(e.Timestamp), nil
-}
-
-// nodeAdded defines a node added event.
+// nodeAdded represents a node added event.
+//
+// Implements history.DataInserter.
 type nodeAdded struct {
-	sqlEvent
+	*pb.TimelineEvent
 }
 
-// newNodeAdded constructs a new node added event.
-func newNodeAdded(timestamp time.Time, node string) nodeAdded {
-	return nodeAdded{
-		sqlEvent: sqlEvent{
-			Timestamp: timestamp,
-			EventType: nodeAddedType,
-			Node:      sql.NullString{String: node, Valid: true},
-		},
-	}
-}
-
-// Insert inserts event into storage using the provided execer.
-func (e nodeAdded) Insert(ctx context.Context, execer history.Execer) error {
+func (r *nodeAdded) Insert(ctx context.Context, execer history.Execer) error {
 	const insertStmt = "INSERT INTO events (timestamp, type, node) VALUES (?,?,?)"
-	args := []interface{}{e.Timestamp, e.EventType, e.Node.String}
+	e := r.GetNodeAdded()
+	if e == nil {
+		return trace.BadParameter("expected NodeAdded, got nil")
+	}
+	args := []interface{}{r.GetTimestamp().ToTime(), history.NodeAdded, e.GetNode()}
 	return trace.Wrap(execer.Exec(ctx, insertStmt, args...))
 }
 
-// ProtoBuf returns event as a protobuf message.
-func (e nodeAdded) ProtoBuf() (*pb.TimelineEvent, error) {
-	return history.NewNodeAdded(e.Timestamp, e.Node.String), nil
-}
-
-// nodeRemoved defines a node added event.
+// nodeRemoved represents a node removed event.
+//
+// Implements history.DataInserter.
 type nodeRemoved struct {
-	sqlEvent
-}
-
-// newNodeRemoved constructs a new node removed event.
-func newNodeRemoved(timestamp time.Time, node string) nodeRemoved {
-	return nodeRemoved{
-		sqlEvent: sqlEvent{
-			Timestamp: timestamp,
-			EventType: nodeRemovedType,
-			Node:      sql.NullString{String: node, Valid: true},
-		},
-	}
+	*pb.TimelineEvent
 }
 
-// Insert inserts event into storage using the provided execer.
-func (e nodeRemoved) Insert(ctx context.Context, execer history.Execer) error {
+func (r *nodeRemoved) Insert(ctx context.Context, execer history.Execer) error {
 	const insertStmt = "INSERT INTO events (timestamp, type, node) VALUES (?,?,?)"
-	args := []interface{}{e.Timestamp, e.EventType, e.Node.String}
+	e := r.GetNodeRemoved()
+	if e == nil {
```

Comment:
This is also not very clear, what's GetNodeRemoved() supposed to return, why can it be nil and what to do in this case? Sorry, commenting on existing code but just thinking this may be confusing if anybody else needs to add a new timeline event for example.

Reply:
d65aa55
`Get<EventType>` does a type assertion and returns the event if ok, or nil o.w.
Replaced the function call and just use a type assertion instead. Should be easier to infer functionality of the code.